### PR TITLE
Add Read Impl for &Stdin

### LINKED ISSUE #95622: `impl Read for &Stdin` is missing
It is possible to write to a `&Stdout` thanks to `impl Write for &Stdout`, but it is not possible to read from a `&Stdin` as the corresponding `impl Read for &Stdin` is missing. This seems to be entirely an accident, given that [`read_line`](https://doc.rust-lang.org/nightly/std/io/struct.Stdin.html#method.read_line) *does* work on `&Stdin`.

Cc @rust-lang/libs-api

### PR DESCRIPTION
**Pull request by Dajamante**
r? @oli-obk
fixes #95622

### CODE REVIEW

**Comment by JohnCSimon** [comment]
triage: 
@m-ou-se this has sat since July - can you post the status of the review?

**Comment by thomcc** [comment]
This needs ACP.

**Comment by Ekleog** [comment]
If it does, then I think https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html?highlight=acp should be updated? The current documentation seems to indicate that ACP are entirely optional, and are only to avoid working on a PR without it eventually landing (the "Note that an ACP is not strictly required" paragraph in particular).

AFAIU that document's suggestions are that this PR should just go through a T-libs FCP, which seems to be confirmed by [this comment in the related issue](https://github.com/rust-lang/rust/issues/95622#issuecomment-1087031086)

File: library/std/src/io/stdio.rs [hunk 1]
```diff
@@ -437,6 +437,28 @@ impl Read for Stdin {
     }
 }
 
+#[stable(feature = "rust1", since = "1.0.0")]
```
Comment (dtolnay) [inline library/std/src/io/stdio.rs:440]: Please choose some other unique feature name and set the stabilization version to "CURRENT_RUSTC_VERSION".

**Review by dtolnay — CHANGES_REQUESTED** [review]

**Comment by dtolnay** [comment]
@rust-lang/libs-api:
@rfcbot fcp merge

Quoting @RalfJung in https://github.com/rust-lang/rust/issues/95622#issue-1191032532:

> It is possible to write to a `&Stdout` thanks to `impl Write for &Stdout`, but it is not possible to read from a `&Stdin` as the corresponding `impl Read for &Stdin` is missing. This seems to be entirely an accident, given that [`read_line`](https://doc.rust-lang.org/nightly/std/io/struct.Stdin.html#method.read_line) *does* work on `&Stdin`.

We implement `std::io::Read` for references already where appropriate, such as for `&File` and `&TcpStream`.

**Comment by dtolnay** [comment]
I filled in a feature name to resolve https://github.com/rust-lang/rust/pull/99153#pullrequestreview-1845022029, and an implementation of `Read::read_buf` which was added in `impl Read for Stdin` by #108326 after this PR was originally opened.

**Review by dtolnay — APPROVED** [review]
Thank you!

**Pull request merged**